src/main/java/com/openshift/internal/restclient/api/capabilities/PodExec.java

```diff
@@ -0,0 +1,151 @@
+
+	private static final Logger LOG = LoggerFactory.getLogger(IPodExec.class);
+	private static final String CAPABILITY = "exec";
+
+	private static final String COMMAND = "command";
+
+	private static final String K8S_PROTOCOL_HEADER = "X-Stream-Protocol-Version";
+	private static final String K8S_PROTOCOL = "channel.k8s.io";
+
+	private final IPod pod;
+	private final DefaultClient client;
+	private final IApiTypeMapper mapper;
+
+	public PodExec(IPod pod, IClient client) {
+		super( pod, client, CAPABILITY );
+		this.pod = pod;
+		this.client = client.adapt(DefaultClient.class);
+		this.mapper = client.adapt(IApiTypeMapper.class);
+	}
+	
+	@Override
+	public String getName() {
+		return PodExec.class.getSimpleName();
+	}
+
+	@Override
+	public IStoppable start( IPodExecOutputListener listener, Options options, String... commands ) {
+
+		Map<String,String> parameters = options.getMap();
+
+		OkHttpClient okClient = client.adapt(OkHttpClient.class);
+
+		URLBuilder urlBuilder = new URLBuilder(client.getBaseURL(), mapper)
+				.resource(pod)
+				.subresource(CAPABILITY)
+				.addParameters(parameters);
+
+		// The main command and all arguments are specified as 'command' parameters
+		for ( String command : commands ) {
+			urlBuilder.addParmeter( COMMAND, command );
+		}
+
+		final String endpoint = urlBuilder.websocket();
+
+		Request request = client.newRequestBuilderTo(endpoint, IHttpConstants.MEDIATYPE_ANY)
+				.method( "GET", null )
+				.addHeader(K8S_PROTOCOL_HEADER, K8S_PROTOCOL )
+				// Unless we mark this as ignored, exceptions triggered by interceptor would be lost in dispatcher thread
+				.tag( new ResponseCodeInterceptor.Ignore() {} )
+				.build();
+
+		WebSocketCall call = WebSocketCall.create(okClient, request);
+		ExecOutputListenerAdapter adapter = new ExecOutputListenerAdapter(call, listener);
+		call.enqueue(adapter);
+		return adapter;
+	}
+
+	static class ExecOutputListenerAdapter extends WebSocketAdapter implements IStoppable{
+
+		private final IPodExecOutputListener listener;
+		private final WebSocketCall call;
+		private AtomicBoolean open = new AtomicBoolean(false);
+
+		public ExecOutputListenerAdapter(WebSocketCall call, IPodExecOutputListener listener) {
+			this.call = call;
+			this.listener = listener;
+		}
+
+		@Override
+		public void stop() {
+			call.cancel();
+		}
+
+		@Override
+		public void onOpen(WebSocket webSocket, Response response) {
+			if(open.compareAndSet(false, true)) {
+				listener.onOpen();
+			}
+		}
+
+		@Override
+		public void onClose(int code, String reason) {
+			if( open.compareAndSet(true, false) ) {
+				listener.onClose(code, reason);
+			}
+		}
+
+		@Override
+		public void onFailure(IOException e, Response response) {
+			listener.onFailure(e);
+		}
+
+		@Override
+		public void onMessage(ResponseBody message) throws IOException {
+			int type = message.byteStream().read();
+			String msg = message.string();
+			switch ( type ) {
```

Comment:
What dictates that type 1, 2, or 3 map to stdout, stderr, execerr ?  If in this pull I'm missing it / please advise.  Otherwise, a comment stating what controls that might be useful for other folks who view this in the future.